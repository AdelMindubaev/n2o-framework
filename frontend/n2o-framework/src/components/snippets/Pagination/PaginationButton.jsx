/**
 * Created by emamoshin on 21.09.2017.
 */
import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

/**
 * Компонент PaginationButton
 * @reactProps {boolean} active - текущая страница
 * @reactProps {boolean} disabled - флаг неактивноти кнопки пагинации
 * @reactProps {boolean} noBorder - флаг включения/выключения бордера
 * @reactProps {function} onSelect - функция клика по кнопке пагинации
 * @reactProps {number} eventKey
 * @reactProps {string|number} label - текст внутри кнопки пагинации
 */
export const PaginationButton = (props) => {
    const { disabled, onSelect, eventKey, label, active, noBorder, tabIndex } = props

    const handleClick = (e) => {
        e.preventDefault()

        if (disabled) {
            return
        }

        if (onSelect) {
            onSelect(eventKey, e)
        }
    }

    return (
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
        <li
            className={classNames('page-item', { active, disabled })}
            onClick={handleClick}
        >
            {/* eslint-disable-next-line jsx-a11y/anchor-is-valid,jsx-a11y/control-has-associated-label */}
            <a
                className={classNames('page-link', noBorder ? 'no-border' : '')}
                href="#"
                tabIndex={tabIndex}
            >
                { label }
            </a>
        </li>
    )
}

PaginationButton.propTypes = {
    active: PropTypes.bool,
    tabIndex: PropTypes.number,
    disabled: PropTypes.bool,
    noBorder: PropTypes.bool,
    onSelect: PropTypes.func,
    eventKey: PropTypes.number,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.element]),
}

export default PaginationButton
