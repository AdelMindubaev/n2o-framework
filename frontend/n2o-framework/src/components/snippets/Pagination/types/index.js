import { shape, number, string, bool, oneOf, func, object } from 'prop-types'

export const pagingType = shape({
    /**
     * Стиль
     * */
    layout: oneOf([
        'bordered',
        'flat',
        'separated',
        'bordered-rounded',
        'flat-rounded',
        'separated-rounded',
    ]),
    /**
     * Место расположения пагинации
     * */
    place: oneOf([
        'topLeft',
        'topRight',
        'bottomLeft',
        'bottomRight',
        'topCenter',
        'bottomCenter',
    ]),
    /**
     * Показать/скрыть кнопку быстрого перехода на предыдущую страницу
     */
    prev: bool,
    /**
     * Вид иконки быстрого перехода на предудущую страницу
     * */
    prevIcon: string,
    /**
     * Текст кнопки 'Назад'
     */
    prevLabel: string,
    /**
     * Показать/скрыть кнопку быстрого перехода на следующую страницу
     */
    next: bool,
    /**
     * Вид иконки быстрого перехода на следующую страницу
     */
    nextIcon: string,
    /**
     * Текст кнопки 'Вперед'
     */
    nextLabel: string,
    /**
     * Показать/скрыть кнопку быстрого перехода на первую страницу
     */
    first: bool,
    /**
     * Вид иконки быстрого перехода на первую страницу
     * */
    firstIcon: string,
    /**
     * Текст кнопки 'First'
     */
    firstLabel: string,
    /**
     * Показать/скрыть кнопку быстрого перехода на последнюю страницу
     */
    last: bool,
    /**
     * Вид иконки быстрого перехода на последнюю страницу
     * */
    lastIcon: string,
    /**
     * Текст кнопки 'Last'
     */
    lastLabel: string,
    /**
     * Показать индикатор общего кол-ва записей
     */
    showCount: bool,
    /**
     * Скрывать компонент, если страница единственная
     */
    showSinglePage: bool,
    /**
     * Максимальное кол-во кнопок перехода между страницами
     */
    maxPages: number,
    /**
     * Общее кол-во записей
     */
    count: number,
    /**
     * Кол-во записей на одной странице
     */
    size: number,
    /**
     * Номер активной страницы
     */
    activePage: number,
    /**
     * Сallback нажатия по кнопке страницы
     */
    onSelect: func,
    /**
     * Класс для списка внутри nav
     */
    className: string,
    t: func,
    style: object,
})
