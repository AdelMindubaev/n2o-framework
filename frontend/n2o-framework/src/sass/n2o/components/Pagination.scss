.pagination {
    &.bordered-rounded {
        .page-item .page-link {
            margin: 0;
        }

        .page-item:first-child .page-link {
            border-top-left-radius: 100px;
            border-bottom-left-radius: 100px;
        }

        .page-item:last-child .page-link {
            border-top-right-radius: 100px;
            border-bottom-right-radius: 100px;
        }
    }
}
