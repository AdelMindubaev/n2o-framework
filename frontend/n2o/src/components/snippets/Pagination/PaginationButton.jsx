/**
 * Created by emamoshin on 21.09.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

/**
 * Компонент PaginationButton
 * @reactProps {boolean} active - текущая страница
 * @reactProps {boolean} disabled - флаг неактивноти кнопки пагинации
 * @reactProps {boolean} noBorder - флаг включения/выключения бордера
 * @reactProps {function} onSelect - функция клика по кнопке пагинации
 * @reactProps {number} eventKey
 * @reactProps {string|number} label - текст внутри кнопки пагинации
 */
class PaginationButton extends React.Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    e.preventDefault();

    const { disabled, onSelect, eventKey } = this.props;

    if (disabled) {
      return;
    }

    if (onSelect) {
      onSelect(eventKey, e);
    }
  }

  render() {
    const { label, active, disabled, noBorder } = this.props;
    return (
      <li className={cx('page-item', { active, disabled })} onClick={this.handleClick}>
        <a className={cx('page-link', noBorder ? 'no-border' : '')} href="#">
          {label}
        </a>
      </li>
    );
  }
}

PaginationButton.propTypes = {
  active: PropTypes.bool,
  disabled: PropTypes.bool,
  noBorder: PropTypes.bool,
  onSelect: PropTypes.func,
  eventKey: PropTypes.number,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

export default PaginationButton;
