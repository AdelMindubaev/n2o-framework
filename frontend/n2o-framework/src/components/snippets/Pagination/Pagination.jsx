import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { withTranslation } from 'react-i18next'

import { PaginationButton } from './PaginationButton'

/**
 * Компонент интерфейса разбивки по страницам
 * @reactProps {boolean} prev - показать/скрыть кнопку быстрого перехода на предыдущую страницу
 * @reactProps {boolean} prevLabel - текс кнопки
 * @reactProps {boolean} next - показать/скрыть кнопку быстрого перехода на следующую страницу
 * @reactProps {boolean} nextLabel - текст кнопки
 * @reactProps {boolean} first - показать/скрыть кнопку быстрого перехода на первую страницу
 * @reactProps {boolean} last - показать/скрыть кнопку быстрого перехода на последнюю страницу
 * @reactProps {boolean} withoutBody - скрыть тело пагинации
 * @reactProps {boolean} showCount - показать индикатор общего кол-ва записей
 * @reactProps {boolean} showSinglePage - показывать компонент, если страница единственная
 * @reactProps {number} maxPages - максимальное кол-во кнопок перехода между страницами
 * @reactProps {number} stepIncrement - шаг дополнительной кнопки (1,2.3 ... 11)
 * @reactProps {number} count - общее кол-во записей
 * @reactProps {number} size - кол-во записей на одной странице
 * @reactProps {number} activePage - номер активной страницы
 * @reactProps {function} onSelect - callback нажатия по кнопке страницы
 * @reactProps {string} className - класс для списка внутри nav
 * @example
 * <Pagination onSelect={this.changePage}
 *             activePage={datasource.page}
 *             count={datasource.count}
 *             size={datasource.size}
 *             maxPages={4}
 *             stepIncrement={10} />
 */
class Pagination extends React.Component {
    /**
     * Рендер тела компонента. Алгоритм автоматически высчитывает страницы до и после текущей
     * @param activePage
     * @param pages
     * @param maxPages
     * @param stepIncrement
     * @param onSelect
     * @returns {Array} - вовзращает список кнопок
     */
    renderBodyPaging = (activePage, pages, maxPages, stepIncrement, onSelect) => {
        const pageButtons = []

        let startPage
        let endPage

        if (maxPages && maxPages < pages) {
            startPage = Math.max(
                Math.min(
                    activePage - Math.floor(maxPages / 2, 10),
                    pages - maxPages + 1,
                ),
                1,
            )
            endPage = startPage + maxPages
        } else {
            startPage = 1
            endPage = maxPages
        }

        if (endPage > pages) {
            endPage = pages
        }

        for (let page = startPage; page <= endPage; ++page) {
            pageButtons.push(
                <PaginationButton
                    key={page}
                    tabIndex={0}
                    eventKey={page}
                    label={page}
                    active={page === activePage}
                    onSelect={onSelect}
                />,
            )
        }

        if (stepIncrement && endPage < pages - 1) {
            pageButtons.push(
                <PaginationButton
                    label="..."
                    tabIndex={-1}
                    key="ellipsisMiddle"
                    noBorder
                    disabled
                />,
            )
            pageButtons.push(
                <PaginationButton
                    tabIndex={0}
                    key={
                        activePage + stepIncrement > pages
                            ? pages
                            : activePage + stepIncrement
                    }
                    eventKey={
                        activePage + stepIncrement > pages
                            ? pages
                            : activePage + stepIncrement
                    }
                    label={
                        activePage + stepIncrement > pages
                            ? pages
                            : activePage + stepIncrement
                    }
                    onSelect={onSelect}
                />,
            )
            if (activePage + stepIncrement < pages) {
                pageButtons.push(
                    <PaginationButton
                        label="..."
                        tabIndex={-1}
                        key="ellipsisLast"
                        noBorder
                        disabled
                    />,
                )
            }
        } else if (stepIncrement && endPage === pages - 1) {
            pageButtons.push(
                <PaginationButton
                    key={pages}
                    eventKey={pages}
                    label={pages}
                    onSelect={onSelect}
                />,
            )
        }

        if (startPage > 1) {
            if (startPage > 2) {
                pageButtons.unshift(
                    <PaginationButton
                        label="..."
                        key="ellipsisFirst"
                        tabIndex={-1}
                        noBorder
                        disabled
                    />,
                )
            }

            pageButtons.unshift(
                <PaginationButton
                    key={1}
                    eventKey={1}
                    label="1"
                    tabIndex={0}
                    onSelect={onSelect}
                />,
            )
        }

        return pageButtons
    }

    /**
     * Базовый рендер компонента
     */
    render() {
        const {
            activePage,
            count,
            size,
            maxPages,
            stepIncrement,
            first,
            last,
            prev,
            next,
            showCount,
            showSinglePage,
            onSelect,
            className,
            withoutBody,
            prevIcon,
            prevLabel,
            nextIcon,
            nextLabel,
            firstIcon,
            firstLabel,
            lastIcon,
            lastLabel,
            t,
        } = this.props
        const pages = Math.ceil(count / size, 10) || 1
        const lastPage = Math.ceil(count / size)

        const getViewButtonPrev = () => {
            const leftArrows = {
                none: '',
                short: '&lsaquo;',
                long: '&xlarr;',
            }

            let label

            if (prevIcon && prevLabel) {
                label = `${leftArrows[prevIcon]}&nbsp;${prevLabel}`
            } else if (prevIcon) {
                label = `${leftArrows[prevIcon]}`
            } else if (prevLabel) {
                label = prevLabel
            }

            return (
                <PaginationButton
                    eventKey={activePage - 1}
                    label={label}
                    disabled={activePage === 1}
                    onSelect={onSelect}
                    tabIndex={0}
                />
            )
        }

        const getViewButtonNext = () => {
            const rightArrows = {
                none: '',
                short: '&rsaquo;',
                long: '&xrarr;',
            }

            let label

            if (nextIcon && nextLabel) {
                label = `${nextLabel}&nbsp;${rightArrows[nextIcon]}`
            } else if (nextIcon) {
                label = `${rightArrows[nextIcon]}`
            } else if (nextLabel) {
                label = nextLabel
            }

            return (
                <PaginationButton
                    eventKey={activePage + 1}
                    label={label}
                    disabled={lastPage === activePage}
                    onSelect={onSelect}
                    tabIndex={0}
                />
            )
        }

        const getFirst = () => {
            const firstArrows = {
                none: '',
                short: '&laquo;',
                long: '&larrb;',
            }

            let label

            if (firstIcon && firstLabel) {
                label = `${firstArrows[firstIcon]}&nbsp;${firstLabel}`
            } else if (firstIcon) {
                label = `${firstArrows[firstIcon]}`
            } else if (firstLabel) {
                label = firstLabel
            }

            return (
                <PaginationButton
                    eventKey={1}
                    label={label}
                    disabled={activePage === 1}
                    onSelect={onSelect}
                    /* eslint-disable-next-line jsx-a11y/tabindex-no-positive */
                    tabIndex={1}
                />
            )
        }

        const getLast = () => {
            const lastArrows = {
                none: '',
                short: '&raquo;',
                long: '&rarrb;',
            }

            let label

            if (lastIcon && lastLabel) {
                label = `${lastLabel}&nbsp;${lastArrows[lastIcon]}`
            } else if (lastIcon) {
                label = `${lastArrows[lastIcon]}`
            } else if (lastLabel) {
                label = lastLabel
            }

            return (
                <PaginationButton
                    eventKey={lastPage}
                    label={label}
                    disabled={lastPage === activePage}
                    onSelect={onSelect}
                    tabIndex={0}
                />
            )
        }

        return (
            <nav
                className="n2o-pagination"
                style={{ display: 'flex', alignItems: 'baseline' }}
            >
                {!showSinglePage && pages === 1 ? null : (
                    <ul className={classNames('pagination', 'd-inline-flex', className)}>
                        {first && getFirst()}
                        {prev && getViewButtonPrev()}
                        {!withoutBody &&
                                this.renderBodyPaging(
                                    activePage,
                                    pages,
                                    maxPages,
                                    stepIncrement,
                                    onSelect,
                                )}
                        {next && getViewButtonNext()}
                        {last && getLast()}
                    </ul>
                )}
                {/* eslint-disable react/jsx-one-expression-per-line */}
                {showCount && (
                    <span
                        className="n2o-pagination-info"
                        style={{
                            paddingLeft: !showSinglePage && pages === 1 ? 0 : '1rem',
                            display: 'inline-flex',
                        }}
                    >
                        {`${t('paginationTotal')} ${count}`}

                                &nbsp;
                        {t('paginationInterval', { postProcess: 'interval', count })}
                    </span>
                )}
            </nav>
        )
    }
}

Pagination.propTypes = {
    /**
     * Показать/скрыть кнопку быстрого перехода на предыдущую страницу
     */
    prev: PropTypes.bool,
    /**
     * Вид иконки быстрого перехода на предудущую страницу
     * */
    prevIcon: PropTypes.string,
    /**
     * Текст кнопки 'Назад'
     */
    prevLabel: PropTypes.string,
    /**
     * Показать/скрыть кнопку быстрого перехода на следующую страницу
     */
    next: PropTypes.bool,
    /**
     * Вид иконки быстрого перехода на следующую страницу
     */
    nextIcon: PropTypes.string,
    /**
     * Текст кнопки 'Вперед'
     */
    nextLabel: PropTypes.string,
    /**
     * Показать/скрыть кнопку быстрого перехода на первую страницу
     */
    first: PropTypes.bool,
    /**
     * Вид иконки быстрого перехода на первую страницу
     * */
    firstIcon: PropTypes.string,
    /**
     * Текст кнопки 'First'
     */
    firstLabel: PropTypes.string,
    /**
     * Показать/скрыть кнопку быстрого перехода на последнюю страницу
     */
    last: PropTypes.bool,
    /**
     * Вид иконки быстрого перехода на последнюю страницу
     * */
    lastIcon: PropTypes.string,
    /**
     * Текст кнопки 'Last'
     */
    lastLabel: PropTypes.string,
    /**
     * Скрыть тело пагинации
     */
    withoutBody: PropTypes.bool,
    /**
     * Показать индикатор общего кол-ва записей
     */
    showCount: PropTypes.bool,
    /**
     * Скрывать компонент, если страница единственная
     */
    showSinglePage: PropTypes.bool,
    /**
     * Максимальное кол-во кнопок перехода между страницами
     */
    maxPages: PropTypes.number,
    /**
     * Шаг дополнительной кнопки (1,2.3 ... 11)
     */
    stepIncrement: PropTypes.number,
    /**
     * Общее кол-во записей
     */
    count: PropTypes.number,
    /**
     * Кол-во записей на одной странице
     */
    size: PropTypes.number,
    /**
     * Номер активной страницы
     */
    activePage: PropTypes.number,
    /**
     * Сallback нажатия по кнопке страницы
     */
    onSelect: PropTypes.func,
    /**
     * Класс для списка внутри nav
     */
    className: PropTypes.string,
    t: PropTypes.func,
}

Pagination.defaultProps = {
    prev: false,
    prevIcon: 'long',
    prevLabel: 'Previous',
    next: false,
    nextIcon: 'long',
    nextLabel: 'Next',
    first: false,
    firstIcon: 'long',
    firstLabel: '',
    last: false,
    lastIcon: 'long',
    lastLabel: '',
    withoutBody: false,
    showCount: true,
    showSinglePage: false,
    maxPages: 4,
    count: 1,
    size: 1,
    activePage: 1,
    t: () => {},
}

export default withTranslation()(Pagination)
