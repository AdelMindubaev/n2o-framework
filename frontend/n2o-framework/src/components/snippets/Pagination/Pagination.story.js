import React from 'react';
import { storiesOf, forceReRender } from '@storybook/react';
import { StateDecorator, Store } from '@sambego/storybook-state';
import Pagination from './Pagination';

const initPages = {
  simple: 1,
  prevNext: 1,
  firstLast: 1,
  combination: 1,
};

const store = new Store({
  page: 1,
  ...initPages,
});

store.subscribe(forceReRender);

const stories = storiesOf('UI Компоненты/Пагинация', module);

stories.addDecorator(StateDecorator(store));
stories
  .add(
    'Компонент',
    () => {
      const props = {
        prev: false,
        next: false,
        first: false,
        last: false,
        lazy: false,
        showCountRecords: true,
        hideSinglePage: true,
        activePage: store.get('page'),
        count: 100,
        size: 10,
        maxButtons: 4,
        stepIncrement: 10,
      };

      return (
        <Pagination
          onSelect={page => {
            store.set({ page });
          }}
          activePage={store.get('page')}
          {...props}
        />
      );
    },
    {
      jsx: { skip: 1 },
      info: {
        text: `
        Компонент 'Пагинация'
        ~~~js
        import Pagination from 'n2o-framework/lib/components/snippets/Pagination/Pagination';
        
        <Pagination
          onSelect={onSelect}
          activePage={activePage}
          prev={false}
          next={false}
          first={false}
          last={false}
          lazy={false}
          showCountRecords
          hideSinglePage
          count={100}
          size={10}
          maxButtons={4}
          stepIncrement={10}
        />
        ~~~
        `,
      },
    }
  )
  .add(
    'Интерактивное использование',
    () => (
      <Pagination
        onSelect={page => {
          store.set({ page });
        }}
        activePage={store.get('page')}
        size={10}
        count={100}
        maxButtons={4}
        stepIncrement={10}
      />
    ),
    {
      jsx: { skip: 1 },
      info: {
        text: `
        Компонент 'Пагинация'
        ~~~js
        import Pagination from 'n2o-framework/lib/components/snippets/Pagination/Pagination';
        
        <Pagination
          onSelect={onSelect}
          activePage={activePage}
          size={10}
          count={100}
          maxButtons={4}
          stepIncrement={10}
        />
        ~~~
        `,
      },
    }
  )
  .add(
    'Вариации',
    () => (
      <React.Fragment>
        <div className="row">
          <Pagination
            activePage={store.get('simple')}
            onSelect={page => store.set({ simple: page })}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('prevNext')}
            onSelect={page => store.set({ prevNext: page })}
            prev={true}
            next={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('firstLast')}
            onSelect={page => store.set({ firstLast: page })}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('combination')}
            onSelect={page => store.set({ combination: page })}
            prev={true}
            next={true}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
      </React.Fragment>
    ),
    {
      jsx: { skip: 1 },
      info: {
        text: `
        Компонент 'Пагинация'
        ~~~js
        import Pagination from 'n2o-framework/lib/components/snippets/Pagination/Pagination';
        
        <Pagination
            activePage={activePage}
            onSelect={onSelect}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
          <Pagination
            activePage={activePage}
            onSelect={onSelect}
            prev={true}
            next={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
          <Pagination
            activePage={activePage}
            onSelect={onSelect}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
          <Pagination
            activePage={activePage}
            onSelect={onSelect}
            prev={true}
            next={true}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        ~~~
        `,
      },
    }
  );
