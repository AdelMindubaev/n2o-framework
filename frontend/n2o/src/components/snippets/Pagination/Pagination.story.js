import React from 'react';
import { storiesOf, forceReRender } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, number } from '@storybook/addon-knobs/react';
import { StateDecorator, Store } from '@sambego/storybook-state';
import Pagination from './Pagination';
import { withState } from '@dump247/storybook-state';

const initPages = {
  simple: 1,
  prevNext: 1,
  firstLast: 1,
  combination: 1,
};

const store = new Store({
  page: 1,
  ...initPages,
});

store.subscribe(forceReRender);

const stories = storiesOf('UI Компоненты/Пагинация', module);

stories.addDecorator(withKnobs);
stories.addDecorator(StateDecorator(store));
stories
  .addWithJSX(
    'Компонент',
    () => {
      const props = {
        prev: boolean('prev', false),
        next: boolean('next', false),
        first: boolean('first', false),
        last: boolean('last', false),
        lazy: boolean('lazy', false),
        showCountRecords: boolean('showCountRecords', true),
        hideSinglePage: boolean('hideSinglePage', true),
        activePage: number('activePage', store.get('page')),
        count: number('count', 100),
        size: number('size', 10),
        maxButtons: number('maxButtons', 4),
        stepIncrement: number('stepIncrement', 10),
      };

      return (
        <Pagination
          onSelect={page => {
            action('pagination-item-click')(page);
            store.set({ page });
          }}
          activePage={store.get('page')}
          {...props}
        />
      );
    },
    { jsx: { skip: 1 } }
  )
  .addWithJSX(
    'Интерактивное использование',
    () => (
      <Pagination
        onSelect={page => {
          store.set({ page });
        }}
        activePage={store.get('page')}
        size={10}
        count={100}
        maxButtons={4}
        stepIncrement={10}
      />
    ),
    { jsx: { skip: 1 } }
  )
  .addWithJSX(
    'Вариации',
    () => (
      <React.Fragment>
        <div className="row">
          <Pagination
            activePage={store.get('simple')}
            onSelect={page => store.set({ simple: page })}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('prevNext')}
            onSelect={page => store.set({ prevNext: page })}
            prev={true}
            next={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('firstLast')}
            onSelect={page => store.set({ firstLast: page })}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
        <div className="row">
          <Pagination
            activePage={store.get('combination')}
            onSelect={page => store.set({ combination: page })}
            prev={true}
            next={true}
            first={true}
            last={true}
            size={10}
            count={100}
            maxButtons={4}
            stepIncrement={10}
          />
        </div>
      </React.Fragment>
    ),
    { jsx: { skip: 1 } }
  );
