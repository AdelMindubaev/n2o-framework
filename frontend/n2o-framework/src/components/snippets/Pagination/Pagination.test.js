import React from 'react';
import Pagination from './Pagination';

const WrapperComp = props => (
  <Pagination {...props} />
);

const setup = (propOverrides, method) => {
  const props = Object.assign(
    {
      // use this to assign some default props
    },
    propOverrides
  );

  let wrapper;

  if (method == 'mount') {
    wrapper = mount(<WrapperComp {...props} />);
  } else {
    wrapper = shallow(<WrapperComp {...props} />);
  }

  return {
    props,
    wrapper,
  };
};

const classicN2oProps = {
  activePage: 1,
  count: 151,
  size: 10,
  maxButtons: 4,
  stepIncrement: 10,
};

describe('<Pagination />', () => {
  it('проверяет рендер с дефолтными настройками', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('проверяет рендер с полным функционалом', () => {
    const { wrapper } = setup({
      prev: true,
      next: true,
      first: true,
      last: true,
      maxButtons: 4,
      stepIncrement: 10,
      count: 151,
      size: 10,
      activePage: 1,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('проверяет рендер с классическими настрйоками n2o', () => {
    const { wrapper } = setup(classicN2oProps);
    expect(wrapper).toMatchSnapshot();
  });

  // it('проверяет состояние на разных страницах', () => {
  //   const values = [3, 4, 5, 6, 7, 13, 14, 15, 16, 1];
  //   const { wrapper } = setup(classicN2oProps, 'mount');
  //   values.forEach(v => {
  //     expect(wrapper).toMatchSnapshot();
  //     wrapper
  //       .find('.page-item')
  //       .at(v - 1)
  //       .simulate('click');
  //     expect(wrapper).toMatchSnapshot();
  //   });
  // });

  it('проверяет работы вызова callback при смене страницы', () => {
    const value = 3;
    const onSelect = jest.fn();
    const { wrapper } = setup(
      { ...classicN2oProps, onSelect: onSelect },
      'mount'
    );

    expect(wrapper).toMatchSnapshot();

    wrapper
      .find('.page-item')
      .at(value - 1)
      .simulate('click');

    expect(onSelect).toHaveBeenCalledWith(3, expect.anything());
  });

  it('проверяет установку класса active для номера страницы', () => {
    const { wrapper } = setup({ ...classicN2oProps }, 'mount');
    expect(wrapper.find('.page-item.active > .page-link').text()).toEqual('1');
    wrapper.setProps({ activePage: 3 });
    wrapper.update();
    expect(wrapper.find('.page-item.active > .page-link').text()).toEqual('3');
  });
});
