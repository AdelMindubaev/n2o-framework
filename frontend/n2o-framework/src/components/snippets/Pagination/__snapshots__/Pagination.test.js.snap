// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> проверяет работы вызова callback при смене страницы 1`] = `
<WrapperComp
  activePage={1}
  count={151}
  maxButtons={4}
  onSelect={[MockFunction]}
  size={10}
  stepIncrement={10}
>
  <withI18nextTranslation(Pagination)
    activePage={1}
    count={151}
    maxButtons={4}
    onSelect={[MockFunction]}
    size={10}
    stepIncrement={10}
  >
    <Pagination
      activePage={1}
      count={151}
      first={false}
      hideSinglePage={true}
      i18n={Object {}}
      last={false}
      maxButtons={4}
      next={false}
      nextText={null}
      onSelect={[MockFunction]}
      prev={false}
      prevText={null}
      showCountRecords={true}
      size={10}
      stepIncrement={10}
      t={[Function]}
      tReady={false}
      withoutBody={false}
    >
      <nav
        className="n2o-pagination"
        style={
          Object {
            "alignItems": "baseline",
            "display": "flex",
          }
        }
      >
        <ul
          className="pagination d-inline-flex"
        >
          <PaginationButton
            active={true}
            eventKey={1}
            key="1"
            label={1}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item active"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 1,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            active={false}
            eventKey={2}
            key="2"
            label={2}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 2,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            active={false}
            eventKey={3}
            key="3"
            label={3}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 3,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            active={false}
            eventKey={4}
            key="4"
            label={4}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 4,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            active={false}
            eventKey={5}
            key="5"
            label={5}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 5,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            disabled={true}
            key="ellipsisMiddle"
            label="..."
            noBorder={true}
            tabIndex={-1}
          >
            <li
              className="page-item disabled"
              onClick={[Function]}
            >
              <a
                className="page-link no-border"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "...",
                  }
                }
                href="#"
                tabIndex={-1}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            eventKey={11}
            key="11"
            label={11}
            onSelect={[MockFunction]}
            tabIndex={0}
          >
            <li
              className="page-item"
              onClick={[Function]}
            >
              <a
                className="page-link"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": 11,
                  }
                }
                href="#"
                tabIndex={0}
              />
            </li>
          </PaginationButton>
          <PaginationButton
            disabled={true}
            key="ellipsisLast"
            label="..."
            noBorder={true}
            tabIndex={-1}
          >
            <li
              className="page-item disabled"
              onClick={[Function]}
            >
              <a
                className="page-link no-border"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "...",
                  }
                }
                href="#"
                tabIndex={-1}
              />
            </li>
          </PaginationButton>
        </ul>
        <span
          className="n2o-pagination-info"
          style={
            Object {
              "display": "inline-flex",
              "paddingLeft": "1rem",
            }
          }
        >
          paginationTotal 151
           
          paginationCount
        </span>
      </nav>
    </Pagination>
  </withI18nextTranslation(Pagination)>
</WrapperComp>
`;

exports[`<Pagination /> проверяет рендер с дефолтными настройками 1`] = `<withI18nextTranslation(Pagination) />`;

exports[`<Pagination /> проверяет рендер с классическими настрйоками n2o 1`] = `
<withI18nextTranslation(Pagination)
  activePage={1}
  count={151}
  maxButtons={4}
  size={10}
  stepIncrement={10}
/>
`;

exports[`<Pagination /> проверяет рендер с полным функционалом 1`] = `
<withI18nextTranslation(Pagination)
  activePage={1}
  count={151}
  first={true}
  last={true}
  maxButtons={4}
  next={true}
  prev={true}
  size={10}
  stepIncrement={10}
/>
`;
